flutter_local_notifications/linux/flutter_local_notifications_plugin.cc

```diff
@@ -0,0 +1,490 @@
+    const auto topLevel = gtk_widget_get_toplevel(view);
+    return topLevel;
+  }
+
+  GtkApplication* getApplication() const {
+    return gtk_window_get_application(GTK_WINDOW(getTopLevel()));
+  }
+
+  std::string getApplicationId() const {
+    GValue id = G_VALUE_INIT;
+    g_object_get_property(G_OBJECT(getApplication()), "application-id", &id);
+    std::string ret = g_value_get_string(&id);
+    g_value_unset(&id);
+    return ret;
+  }
+
+  FlMethodResponse* initialize(FlMethodCall* call) {
+    const auto args = fl_method_call_get_args(call);
+    if (args && fl_value_get_type(args) == FL_VALUE_TYPE_MAP) {
+      const auto defaultIconValue = fl_value_lookup_string(args, "defaultIcon");
+      if (defaultIconValue && fl_value_get_type(defaultIconValue) == FL_VALUE_TYPE_MAP) {
+        default_icon = CreateIconFromFlValue(defaultIconValue);
+      }
+    }
+
+    const auto app = getApplication();
+    const GActionEntry actionEntries[] = {
+      {
+        NotificationActionName,
+        [](GSimpleAction* action, GVariant* param, gpointer opaque) {
+          g_autoptr(GVariant) id = g_variant_get_child_value(param, 0);
+          const auto idValue = g_variant_get_int64(id);
+          g_autoptr(GVariant) payload = g_variant_get_child_value(param, 1);
+          const auto payloadValue = g_variant_get_string(payload, nullptr);
+
+          const auto plugin = static_cast<FlutterLocalNotificationsPlugin*>(opaque);
+
+          g_autoptr(FlValue) arg = fl_value_new_map();
+          fl_value_set_string_take(arg, "id", fl_value_new_int(idValue));
+          fl_value_set_string_take(arg, "payload", fl_value_new_string(payloadValue));
+          fl_method_channel_invoke_method(plugin->channel, "selectNotification", arg, nullptr, nullptr, nullptr);
+        },
+        "(xs)"
+      },
+      {
+        NotificationButtonActionName,
+        [](GSimpleAction* action, GVariant* param, gpointer opaque) {
+          g_autoptr(GVariant) id = g_variant_get_child_value(param, 0);
+          const auto idValue = g_variant_get_int64(id);
+          g_autoptr(GVariant) buttonId = g_variant_get_child_value(param, 1);
+          const auto buttonIdValue = g_variant_get_string(buttonId, nullptr);
+
+          const auto plugin = static_cast<FlutterLocalNotificationsPlugin*>(opaque);
+
+          g_autoptr(FlValue) arg = fl_value_new_map();
+          fl_value_set_string_take(arg, "id", fl_value_new_int(idValue));
+          fl_value_set_string_take(arg, "buttonId", fl_value_new_string(buttonIdValue));
+          fl_method_channel_invoke_method(plugin->channel, "selectNotificationButton", arg, nullptr, nullptr, nullptr);
+        },
+        "(xs)"
+      }
+    };
+    g_action_map_add_action_entries(G_ACTION_MAP(app), actionEntries, std::size(actionEntries), this);
+    return FL_METHOD_RESPONSE(fl_method_success_response_new(nullptr));
+  }
+
+  FlMethodResponse* show(FlMethodCall* call) {
```

Comment:
this is probably the whole plugin's most important function, but it's a bit hard to follow because it's so long and does many things. would it be better to try to split it up into smaller chunks?